### DISCUSSION TITLE
file not existing frustration

### DISCUSSION THREAD

**Discussion by endremborza** [nushell/nushell#13237] [General]
Hello.

After spending over an hour setting up custom completions, I ran into this, directly copied message which I think just about put me over the edge to give up on nu in its current form:

```
  × File not found
    ╭─[/home/borza/.config/nushell/env.nu:50:12]
 49 │ if ($CUSTOM_COMPLETIONS_FILE | path exists) {
 50 │     source $CUSTOM_COMPLETIONS_FILE
    ·            ────────────┬───────────
    ·                        ╰── File not found: /home/borza/setup-repos/nu_scripts/custom-completions/use.nu
 51 │ }
    ╰────
  help: sourced files need to be available before your script is run
```

I really should not need to put anything else here, there are no lines between the two, how is this possible?

its 0.94.2

I did a simpler one for good measure

```
Error: nu::parser::sourced_file_not_found

  × File not found
    ╭─[/home/borza/.config/nushell/env.nu:50:12]
 49 │ if ("any.nu" | path expand | path exists) {
 50 │     source ("any.nu" | path expand)
    ·            ────────────┬───────────
    ·                        ╰── File not found: /home/borza/any.nu
 51 │ } 
    ╰────
  help: sourced files need to be available before your script is run
```

**Comment by fdncred** [comment]
I believe this is working as expected. You can read more about how nushell works in https://www.nushell.sh/book/thinking_in_nu.html and https://www.nushell.sh/book/how_nushell_code_gets_run.html. Nushell is scoped so even if you can get a file to source within and `if` block, it will probably only be visible to that block and not outside it.

Example:
```nushell
> open test.nu
def test_fun [] {
  print "hello from test_fun custom command"
}
> # run test_fun inside the scope
> if true { source /home/fdncred/src/test.nu; test_fun }
hello from test_fun custom command
> # run test_fun outside the scope
> test_fun
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #15:1:1]
 1 │ test_fun
   · ────┬───
   ·     ╰── Command `test_fun` not found
   ╰────
  help: `test_fun` is neither a Nushell built-in or a known external command
```

**Comment by endremborza** [comment]
I understand.

I would not say "as expected", but I accept that for some reason, for nushell this is ok. i think scoped ifs for the language of a shell are ludicrous, but I accept it as a design choice. what I think is not a matter of opinion is that this error message for this script looks embarrassing. why allow for if blocks (or certain commands in if blocks) if they do not matter at all? this "help" helps nobody.

for a language to throw an error of a file not existing where the previous line explicitly checks for that file existing, I think needs a bit more immediate lucid explanation, than it "needs to ve available"

**Reply by Lockszmith-GH** [reply]
well the error is sort of a 'compilation error', but I'm with you - I'm new to the language - and dynamically loading settings into my shell is something I do with older, more primitive shells.
This is one of the reasons I'm holding back on nu - because I want to make things dynamic, based on my environment.

**Comment by fdncred** [comment]
I think the typical pattern for sourcing/using things is to construct a const variable based on a condition. Here's a quick example.
```nushell
❯ const CHEAT = ("cheat-sheet.nu" | path expand)
❯ const NOCHEAT = 'C:\Users\fdncred\source\repos\no-cheat-sheet.nu'
❯ const file_to_source = (if ($CHEAT | path exists) { $CHEAT } else { $NOCHEAT })
❯ source $file_to_source
```
So, `use` and `source` can work if the "variables" are constants and not scoped.

**Reply by Lockszmith-GH** [reply]
I wonder if `/dev/null` can be used here as the 'NOCHEAT'. Not as convenient as bash, but at least a dynamic solution.
This would make things so much more convenient on the interactive side.

Not at my PC right now, I'll post here if successful after I'll test it out.
